Content.Server/DeltaV/Abilities/Psionics/PrecognitionPowerSystem.cs

```diff
@@ -0,0 +1,222 @@
+    /// Gets the precognition result message corosponding to the passed event id.
+    /// </summary>
+    /// <returns>message string corosponding to the event id passed</returns>
+    private string GetResultMessage(EntProtoId? eventId)
+    {
+        foreach(var (eventProto, precognitionResult) in AllPrecognitionResults())
+            if (eventProto.ID == eventId && precognitionResult != null)
+                return precognitionResult.Message;
+        Log.Warning("Prototype " + eventId + "does not have an associated precognitionResult!");
+        return string.Empty;
+    }
+
+    /// <summary>
+    /// </summary>
+    /// <returns>The localized string of a weighted randomly chosen precognition result</returns>
+    public string? GetRandomResult()
+    {
+        var precognitionResults = AllPrecognitionResults();
+        var sumOfWeights = 0;
+        foreach (var precognitionResult in precognitionResults.Values)
+            sumOfWeights += (int)precognitionResult.Weight;
+
+        sumOfWeights = _random.Next(sumOfWeights);
+        foreach (var precognitionResult in precognitionResults.Values)
+        {
+            sumOfWeights -= (int)precognitionResult.Weight;
+
+            if (sumOfWeights <= 0)
+                return precognitionResult.Message;
+        }
+
+        Log.Error("Result was not found after weighted pick process!");
+        return null;
+    }
+
+    /// <summary>
+    /// Gets the soonest nextEvent to occur within the window.
+    /// </summary>
+    /// <param name="minDetectWindow"></param> The earliest reletive time that will be return a nextEvent
+    /// <param name="maxDetectWindow"></param> The latest reletive latest time that will be return a nextEvent
+    /// <param name="earliestNextEvent"></param> The next nextEvent to occur within the window
+    /// <returns>True if event was found in timeframe false otherwise.</returns>
+    private bool TryFindEarliestNextEvent(TimeSpan minDetectWindow, TimeSpan maxDetectWindow, out NextEventComponent? earliestNextEvent)
+    {
+        TimeSpan? earliestNextEventTime = null;
+        earliestNextEvent = null;
+        var query = EntityQueryEnumerator<NextEventComponent>();
+        while (query.MoveNext(out _, out var nextEventComponent))
+        {
+            // Update if the event is the most recent event that isnt too close or too far from happening to be of use
+            if (nextEventComponent.NextEventTime > _gameTicker.RoundDuration() + minDetectWindow
+                && nextEventComponent.NextEventTime < _gameTicker.RoundDuration() + maxDetectWindow
+                && earliestNextEvent == null
+                || nextEventComponent.NextEventTime < earliestNextEventTime)
+                earliestNextEvent ??= nextEventComponent;
+        }
+        if (earliestNextEvent == null)
+            return false;
+        return true;
+    }
+
+    public Dictionary<EntityPrototype, PrecognitionResultComponent> AllPrecognitionResults()
```

Comment:
cache this and update on prototype reload if entityprototype is modified

Reply:
Would it be preferable to make a AllResults Dict in the PrecognitionPowerComponent? And then have it populate in the OnMapInit?

Reply:
Also does it need to update? There isn't behavior for adding new result messages so I don't really see a usecase.